src/trie_tree.rs

```diff
@@ -0,0 +1,456 @@
+            store,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Merkle root
+    pub fn root(&self) -> &H256 {
+        &self.root
+    }
+
+    /// Check empty of the tree
+    pub fn is_empty(&self) -> bool {
+        self.root.is_zero()
+    }
+
+    /// Destroy current tree and retake store
+    pub fn take_store(self) -> S {
+        self.store
+    }
+
+    /// Get backend store
+    pub fn store(&self) -> &S {
+        &self.store
+    }
+
+    /// Get mutable backend store
+    pub fn store_mut(&mut self) -> &mut S {
+        &mut self.store
+    }
+}
+
+impl<H: Hasher + Default, V, S: StoreReadOps<V>> SparseMerkleTree<H, V, S> {
+    /// Build a merkle tree from store, the root will be calculated automatically
+    pub fn new_with_store(store: S) -> Result<SparseMerkleTree<H, V, S>> {
+        let root_branch_key = BranchKey::new(core::u8::MAX, H256::zero());
+        store
+            .get_branch(&root_branch_key)
+            .map(|branch_node| {
+                branch_node
+                    .map(|n| {
+                        merge::<H>(core::u8::MAX, &H256::zero(), &n.left, &n.right).hash::<H>()
+                    })
+                    .unwrap_or_default()
+            })
+            .map(|root| SparseMerkleTree::new(root, store))
+    }
+}
+
+impl<H: Hasher + Default, V: Value, S: StoreReadOps<V> + StoreWriteOps<V>>
+    SparseMerkleTree<H, V, S>
+{
+    /// Update a leaf, return new merkle root
+    /// set to zero value to delete a key
+    pub fn update(&mut self, key: H256, value: V) -> Result<&H256> {
+        // compute and store new leaf
+        let node = MergeValue::from_h256(value.to_h256());
+        // notice when value is zero the leaf is deleted, so we do not need to store it
+        if !node.is_zero() {
+            self.store.insert_leaf(key, value)?;
+        } else {
+            self.store.remove_leaf(&key)?;
+        }
+
+        let mut last_height = core::u8::MAX;
+        loop {
+            // walk from top to bottom
+            let node_key = key.parent_path(last_height);
+            let branch_key = BranchKey::new(last_height, node_key); // this represents a position in the tree
+            if let Some(branch) = self.store.get_branch(&branch_key)? {
+                // if we we found a record in here
+                // we need to determine whether is it a shortcut
+                let (target, another) = if key.is_right(last_height) {
+                    (branch.right, branch.left)
+                } else {
+                    (branch.left, branch.right)
+                };
+
+                match target {
+                    MergeValue::ShortCut {
+                        key: this_key,
+                        value,
+                        height: h,
+                    } => {
+                        if this_key.eq(&key) {
+                            // we update its value
+                            let target = MergeValue::shortcut(key, node.hash::<H>(), h);
```

Comment:
```suggestion
                            let target = MergeValue::shortcut(key, value.to_h256(), h);
```
Would be more readable.